File: video/ai-extensions/video-63-leetcode-5/leetcode5.dart
class Solution {
  String longestPalindrome(String s) {
    if (s.isEmpty) {
      return "";
    }

    int start = 0;
    int end = 0;

    for (int i = 0; i < s.length; i++) {
      int len1 = _expandAroundCenter(s, i, i);
      int len2 = _expandAroundCenter(s, i, i + 1);
      int len = len1 > len2 ? len1 : len2;

      if (len > end - start) {
        start = i - ((len - 1) >> 1);
        end = i + (len >> 1);
      }
    }

    return s.substring(start, end + 1);
  }

  int _expandAroundCenter(String s, int left, int right) {
    while (left >= 0 && right < s.length && s[left] == s[right]) {
      left--;
      right++;
    }
    return right - left - 1;
  }
}

File: video/ai-extensions/video-63-leetcode-5/leetcode5.kt
class Solution {
    fun longestPalindrome(s: String): String {
        if (s.isEmpty()) return ""

        var start = 0
        var end = 0

        for (i in s.indices) {
            val len1 = expandAroundCenter(s, i, i)
            val len2 = expandAroundCenter(s, i, i + 1)
            val len = maxOf(len1, len2)
            if (len > end - start) {
                start = i - (len - 1) / 2
                end = i + len / 2
            }
        }

        return s.substring(start, end + 1)
    }

    private fun expandAroundCenter(s: String, left: Int, right: Int): Int {
        var L = left
        var R = right
        while (L >= 0 && R < s.length && s[L] == s[R]) {
            L--
            R++
        }
        return R - L - 1
    }
}
